File: doc/spec/proposals/reindex.pl
#!/usr/bin/perl -w
# Copyright 2007 Nick Mathewson.  See LICENSE for licensing information.

use strict;

my $propdir = ".";
local *DIR;
local *F;

opendir(DIR, $propdir) || die "Can't open $propdir";
my @files = sort grep { /^\d\d\d-.*[^\~]$/ } readdir(DIR);
closedir DIR;

my %title = ();
my %status = ();

my @KNOWN_STATUSES = qw{
    DRAFT OPEN NEEDS-REVISION NEEDS-RESEARCH ACCEPTED META FINISHED CLOSED
    SUPERSEDED DEAD};

for my $f (@files) {
    my $num = substr($f, 0, 3);
    my $status = undef;
    my $title = undef;
    my $implemented_in = undef;
    my $target = undef;
    my $alleged_fname = undef;
    if ($f !~ /\.txt/) { print "$f doesn't end with .txt\n"; }
    open(F, "$f");
    while (<F>) {
	last if (/^\s*$/);
	if (/^Status: (.*)/) {
	    $status = uc $1;
	    chomp $status;
	}
	if (/^Filename: (.*)/) {
	    $alleged_fname = $1;
	    chomp $alleged_fname;
	}
	if (/^Title: (.*)/) {
	    $title = $1;
	    $title =~ s/\.$//;
	    chomp $title;
	}
	if (/^Implemented-In: (.*)/) {
	    $implemented_in = $1;
	    chomp $implemented_in;
	}
	if (/^Target: (.*)/) {
	    $target = $1;
	    chomp $target;
	}
    }
    close F;
    die "Proposal $num has no status line" if (!defined $status);
    die "I've never heard of status $status in proposal $num"
	unless (grep(/$status/, @KNOWN_STATUSES) == 1);
    die "Proposal $num has no title line" if (!defined $title);
    die "Proposal $num has no Filename line" unless (defined $alleged_fname);
    die "Proposal $num says its fname is $alleged_fname, but it's really $f"
	if ($alleged_fname ne $f);
    print "No Target for proposal $num\n" if (($status eq 'OPEN' or
					       $status eq 'ACCEPTED')
					      and !defined $target);
    print "No Implemented-In for proposal $num\n"
	if (($status eq 'CLOSED' or $status eq 'FINISHED')
	    and !defined $implemented_in);

    $title{$num} = $title;
    $status{$num} = $status;
}

local *OUT;
open(OUT, ">000-index.txt.tmp");

open(F, "000-index.txt") or die "Can't open index file.";
while (<F>) {
    print OUT;
    last if (/^={3,}/);
}
close(F);

print OUT "Proposals by number:\n\n";

for my $num (sort keys %title) {
    print OUT "$num  $title{$num} [$status{$num}]\n";
}

print OUT "\n\nProposals by status:\n\n";
for my $status (@KNOWN_STATUSES) {
    print OUT " $status:\n";
    for my $num (sort keys %status) {
	next unless ($status{$num} eq $status);
	print OUT "   $num  $title{$num}\n";
    }
}

rename('000-index.txt.tmp', '000-index.txt');

File: doc/spec/proposals/reindex.py
#!/usr/bin/python

import re, os
class Error(Exception): pass

STATUSES = """DRAFT NEEDS-REVISION NEEDS-RESEARCH OPEN ACCEPTED META FINISHED
   CLOSED SUPERSEDED DEAD REJECTED""".split()
REQUIRED_FIELDS = [ "Filename", "Status", "Title" ]
CONDITIONAL_FIELDS = { "OPEN" : [ "Target" ],
                       "ACCEPTED" : [ "Target "],
                       "CLOSED" : [ "Implemented-In" ],
                       "FINISHED" : [ "Implemented-In" ] }
FNAME_RE = re.compile(r'^(\d\d\d)-.*[^\~]$')
DIR = "."
OUTFILE = "000-index.txt"
TMPFILE = OUTFILE+".tmp"

def indexed(seq):
    n = 0
    for i in seq:
        yield n, i
        n += 1

def readProposal(fn):
    fields = { }
    f = open(fn, 'r')
    lastField = None
    try:
        for lineno, line in indexed(f):
            line = line.rstrip()
            if not line:
                return fields
            if line[0].isspace():
                fields[lastField] += " %s"%(line.strip())
            else:
                parts = line.split(":", 1)
                if len(parts) != 2:
                    raise Error("%s:%s:  Neither field nor continuation"%
                                (fn,lineno))
                else:
                    fields[parts[0]] = parts[1].strip()
                    lastField = parts[0]

        return fields
    finally:
        f.close()

def checkProposal(fn, fields):
    status = fields.get("Status")
    need_fields = REQUIRED_FIELDS + CONDITIONAL_FIELDS.get(status, [])
    for f in need_fields:
        if not fields.has_key(f):
            raise Error("%s has no %s field"%(fn, f))
    if fn != fields['Filename']:
        print `fn`, `fields['Filename']`
        raise Error("Mismatched Filename field in %s"%fn)
    if fields['Title'][-1] == '.':
        fields['Title'] = fields['Title'][:-1]

    status = fields['Status'] = status.upper()
    if status not in STATUSES:
        raise Error("I've never heard of status %s in %s"%(status,fn))
    if status in [ "SUPERSEDED", "DEAD" ]:
        for f in [ 'Implemented-In', 'Target' ]:
            if fields.has_key(f): del fields[f]

def readProposals():
    res = []
    for fn in os.listdir(DIR):
        m = FNAME_RE.match(fn)
        if not m: continue
        if not fn.endswith(".txt"):
            raise Error("%s doesn't end with .txt"%fn)
        num = m.group(1)
        fields = readProposal(fn)
        checkProposal(fn, fields)
        fields['num'] = num
        res.append(fields)
    return res

def writeIndexFile(proposals):
    proposals.sort(key=lambda f:f['num'])
    seenStatuses = set()
    for p in proposals:
        seenStatuses.add(p['Status'])

    out = open(TMPFILE, 'w')
    inf = open(OUTFILE, 'r')
    for line in inf:
        out.write(line)
        if line.startswith("====="): break
    inf.close()

    out.write("Proposals by number:\n\n")
    for prop in proposals:
        out.write("%(num)s  %(Title)s [%(Status)s]\n"%prop)
    out.write("\n\nProposals by status:\n\n")
    for s in STATUSES:
        if s not in seenStatuses: continue
        out.write(" %s:\n"%s)
        for prop in proposals:
            if s == prop['Status']:
                out.write("   %(num)s  %(Title)s"%prop)
                if prop.has_key('Target'):
                    out.write(" [for %(Target)s]"%prop)
                if prop.has_key('Implemented-In'):
                    out.write(" [in %(Implemented-In)s]"%prop)
                out.write("\n")
    out.close()
    os.rename(TMPFILE, OUTFILE)

try:
    os.unlink(TMPFILE)
except OSError:
    pass

writeIndexFile(readProposals())
